controller/src/main.rs

```diff
@@ -1,3 +1,226 @@
+    fn all_nodes(&self) -> Vec<BottlerocketNode> {
+        self.brn_reader.state()
+    }
+
+    /// Returns the set of BottlerocketNode objects which is currently being acted upon.
+    ///
+    /// Nodes are being acted upon if they are not in the `WaitingForUpdate` state, or if their desired state does
+    /// not match their current state.
+    fn active_node_set(&self) -> BTreeMap<String, BottlerocketNode> {
+        self.all_nodes()
+            .iter()
+            .filter(|brn| {
+                brn.status.as_ref().map_or(false, |brn_status| {
+                    brn_status.current_state != BottlerocketNodeState::WaitingForUpdate
+                        || brn_status.current_state != brn.spec.state
+                })
+            })
+            // kube-rs doesn't implement Ord or Hash on ObjectMeta, so we store these in a map indexed by name.
+            // (which are unique within a namespace). `name()` is guaranteed not to panic, as these nodes are all populated
+            // by our `reflector`.
+            .map(|brn| (brn.name(), brn.clone()))
+            .collect()
+    }
+
+    /// Determines next actions for a BottlerocketNode.
+    async fn progress_node(&self, node: BottlerocketNode) -> Result<()> {
+        event!(
+            Level::TRACE,
+            ?node,
+            "Attempting to progress BottlerocketNode."
+        );
+        let node_status = node
+            .status
+            .as_ref()
+            .ok_or(error::Error::NodeWithoutStatus {
+                node_name: node.name().to_string(),
+            })?;
+
+        if node_status.current_state == node.spec.state {
+            // If the node has reached its desired status, it is ready to move on.
+            let desired_spec = node
+                .determine_next_spec()
+                .context(error::NodeSpecCannotBeDetermined)?;
+
+            event!(
+                Level::TRACE,
+                ?node,
+                ?desired_spec,
+                "BottlerocketNode has reached desired status. Modifying spec."
+            );
+
+            self.node_client
+                .update_node_spec(
+                    &node.selector().context(error::NodeSelectorCreation)?,
+                    &desired_spec,
+                )
+                .await
+                .context(error::UpdateNodeSpec)
+        } else {
+            // Otherwise, we need to ensure that the node is making progress in a timely fashion.
+
+            // TODO(seankell) Timeout handling will be added in a future PR.
+            Ok(())
+        }
+    }
+
+    /// This function searches all `BottlerocketNode`s for those which can be transitioned to a new state.
+    /// The state transition is then attempted. If successful, this node should be detected as part of the active
+    /// set during the next iteration of the controller's event loop.
+    async fn search_for_node_to_update(&self) -> Option<BottlerocketNode> {
```

Comment:
Question here. Does this function only find one node to update every time? why not finding all nodes and return a list of nodes which are qualified to update, then main loop can handle them.